### PR TITLE
Store card id and feature flag to enable linking/draining

### PR DESCRIPTION
**Pull request by husobee**
### Summary

1. Adds a flag "ENABLE_LINKING_DRAINING" which when set to false will disable linking of wallets and draining of grants.
2. Stores the card id (tx.destination) as provider_id in the database
3. Draining will transfer to provider_id (tx.destination) instead of anon address.

### Type of change ( select one )

- [ ] Product feature
- [x] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before submitting this PR:

- [ ] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [ ] Is there a clear title that explains what the PR does?
- [ ] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security / privacy review if needed
- [ ] Have you performed a self review of this PR?

### CODE REVIEW

File: cmd/wallets.go [hunk 1]
```diff
@@ -52,6 +53,12 @@ func init() {
 	must(viper.BindPFlag("wallets-in-migration-flag", walletsCmd.PersistentFlags().Lookup("wallets-in-migration-flag")))
 	must(viper.BindEnv("wallets-in-migration-flag", "WALLETS_IN_MIGRATION"))
 
+	// ENABLE_LINKING_DRAINING - enable ability to link wallets and drain wallets
+	walletsCmd.PersistentFlags().BoolVarP(&enableLinkingDrainingFlag, "enable-linking-draining-flag", "", false,
+		"the in-migration flag disabling the wallets link feature")
```
Comment (evq) [inline cmd/wallets.go:58]: small nit - may want to tweak this, seems the same as above

Comment (husobee) [reply cmd/wallets.go:51]: https://github.com/brave-intl/bat-go/pull/469/commits/2a56be2b9f36338aab261e722af4d8e3f9a4c7dc

File: migrations/0019_wallet_card_id.down.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+ALTER TABLE wallets DROP COLUMN user_deposit_card_id;
```
Comment (evq) [inline migrations/0019_wallet_card_id.down.sql:1]: `card_id` is pretty uphold specific. should we use something like `user_deposit_destination`? or similar

File: cmd/wallets_rest_run.go [hunk 3]
```diff
@@ -58,7 +58,7 @@ func SetupWalletService(ctx context.Context, r *chi.Mux) (*chi.Mux, context.Cont
 				"CreateBraveWallet", wallet.CreateBraveWalletV3))
 
 			// if wallets are being migrated we do not want to over claim, we might go over the limit
-			if !viper.GetBool("wallets-in-migration-flag") {
+			if !viper.GetBool("wallets-in-migration-flag") && viper.GetBool("enable-linking-draining-flag") {
```
[the remark this answers is not in the document]
Comment (husobee) [reply cmd/wallets_rest_run.go]: https://github.com/brave-intl/bat-go/pull/469/commits/3b24f7861df843e71c37ca657745785d6b2730c5

File: promotion/drain.go [hunk 4]
```diff
@@ -96,28 +98,35 @@ func (service *Service) RedeemAndTransferFunds(ctx context.Context, credentials
...
-			service.drainChannel <- tx
-		}
+	enableLinkingDraining, err := strconv.ParseBool(os.Getenv("ENABLE_LINKING_DRAINING"))
+	if err != nil {
+		return nil, fmt.Errorf("invalid enable_linking_draining flag: %w", err)
+	}
 
-		return tx, err
+	if enableLinkingDraining {
+		// first use the provider id, if not exists use anon address
```
[the remark this answers is not in the document]
Comment (husobee) [reply promotion/drain.go]: https://github.com/brave-intl/bat-go/pull/469/commits/3b24f7861df843e71c37ca657745785d6b2730c5

File: promotion/drain.go [hunk 5]
```diff
@@ -96,28 +98,35 @@ func (service *Service) RedeemAndTransferFunds(ctx context.Context, credentials
...
+			if err != nil {
+				return nil, err
+			}
+			if service.drainChannel != nil {
+				service.drainChannel <- tx
+			}
+
+			return tx, err
+		}
+		return nil, errors.New("no anonymous address for wallet for transfer")
```
[the remark this answers is not in the document]
Comment (husobee) [reply promotion/drain.go]: https://github.com/brave-intl/bat-go/pull/469/commits/3b24f7861df843e71c37ca657745785d6b2730c5

File: migrations/0019_wallet_card_id.down.sql (hunk 2 above)
Comment (husobee) [reply migrations/0019_wallet_card_id.down.sql]: https://github.com/brave-intl/bat-go/pull/469/commits/3b24f7861df843e71c37ca657745785d6b2730c5

File: promotion/drain.go [hunk 6]
```diff
@@ -96,28 +98,35 @@ func (service *Service) RedeemAndTransferFunds(ctx context.Context, credentials
...
-		if service.drainChannel != nil {
-			service.drainChannel <- tx
-		}
+	enableLinkingDraining, err := strconv.ParseBool(os.Getenv("ENABLE_LINKING_DRAINING"))
+	if err != nil {
+		return nil, fmt.Errorf("invalid enable_linking_draining flag: %w", err)
+	}
 
-		return tx, err
+	if enableLinkingDraining {
```
[the remark this answers is not in the document]
Comment (husobee) [reply promotion/drain.go]: in https://github.com/brave-intl/bat-go/pull/469/commits/3b24f7861df843e71c37ca657745785d6b2730c5 I am wrapping the route definition for the suggestion drain endpoint.  I can make a custom error if a 404 doesn't work in this case, but that is the same as we are doing for wallet linking if disabled.

File: promotion/controllers_test.go [hunk 7]
```diff
@@ -1292,8 +1293,7 @@ func (suite *ControllersTestSuite) TestSuggestionDrain() {
 	err = s.Sign(privKey, crypto.Hash(0), req)
 	suite.Require().NoError(err)
 
-	anonymousAddress := uuid.Must(uuid.FromString(w.ProviderID))
-	info.AnonymousAddress = &anonymousAddress
+	info.UserDepositDestination = w.ProviderID
```
Comment (corymcdonald) [inline promotion/controllers_test.go:1296]: This code seems like the same as before, do we still want to validate the `uuid.Must` property of the ProviderID?

Will this parameter change with the client change? It looks like we're mapping the existing anonymous addresses to `UserDepositDestination`?

File: wallet/service.go [hunk 8]
```diff
@@ -142,11 +142,8 @@ func (service *Service) LinkWallet(
 			return handlers.WrapError(errors.New("wallets do not match"), "unable to match wallets", http.StatusForbidden)
 		}
 	} else {
-		destination := info.ProviderID
-		if info.ProviderID == "" {
-			destination = tx.Destination
-		}
-		err := service.Datastore.LinkWallet(info.ID, destination, providerLinkingID, anonymousAddress, depositProvider)
+		// tx.Destination will be stored as "ProviderID"
```
Comment (corymcdonald) [inline wallet/service.go:145]: Is this comment still true? It looks like we're inserting into the db as `user_deposit_destination` but there's not an associated migration. A case of bad find/replace?

File: promotion/controllers_test.go (hunk 7 above)
Comment (husobee) [reply promotion/controllers_test.go:1296]: UserDepositDestination is a string.  Current use cases happen to use a uuid, but that might not always be the case for wallet deposit providers.  In this test we are just filling in a value to make sure the value is right.

We are not directly mapping from anonymous address to destination, they are two seperate values.  The anonymous address is a uuid that the browser creates and we give uphold to obfuscate the actual destination.   Destination is the actual destination for the deposit.

I agree that this test is a bit confusing, it should be more clear that anonymous address and this deposit destination are different things.

File: wallet/service.go (hunk 8 above)
Comment (husobee) [reply wallet/service.go:145]: will correct.  thanks.

Comment (husobee) [reply wallet/service.go]: https://github.com/brave-intl/bat-go/pull/469/commits/32ffbbfdd0dc32ff53aa77414a0d9d7cd24b276c

File: cmd/wallets.go [hunk 9]
```diff
@@ -20,10 +20,10 @@ var (
 		Short: "provides REST api services",
 		Run:   WalletRestRun,
 	}
-	db                     string
-	walletsFeatureFlag     bool
-	walletsInMigrationFlag bool
-	roDB                   string
+	db                        string
+	walletsFeatureFlag        bool
+	enableLinkingDrainingFlag bool
```
[the remark this answers is not in the document]
Comment (husobee) [reply cmd/wallets.go]: https://github.com/brave-intl/bat-go/pull/469/commits/32ffbbfdd0dc32ff53aa77414a0d9d7cd24b276c

**Pull request merged**